include/swift/Parse/Parser.h

```diff
@@ -1782,7 +1783,8 @@ class Parser {
   /// This function should be called on a Parser set up on the command line
   /// argument code.
   ParserStatus
-  parseAvailabilityMacroDefinition(AvailabilityMacroDefinition &Result);
+  parseAvailabilityMacroDefinition(AvailabilityMacroDefinition &Result,
+                                   bool isUnavailability = false);
```

Comment:
An availability macro should serve the dual purpose for working both with `#available` and `#unavailable`. The availability spec list for an availability macro behaves like the spec list for `#unavailable` (see usage of `@available(SwiftStdlib 5.5, *)` in the stdlib; the definition of SwiftStdlib from compiler flags doesn't include `*`). That said, there's no fix-it in that case. Instead of having a pair of booleans (`isUnavailability` and isParsingMacroDefinition`) in multiple places, which indicates four possibilities, I think it would be best to try to use a dedicated enum with 3 cases. Something like:


```
enum class AvailabilitySpecSource: uint8_t {
  Available,
  Unavailable,
  Macro
}

void printAvailabilitySpecSource(llvm::raw_ostream &os, AvailabilitySpecSource source) { ... }
void dumpAvailabilitySpecSource(AvailabilitySpecSource source) { /* call print with llvm::errs() */ }
```

Then the body of this function would pass the `Macro` case to `parseAvailabilitySpecList`. Other places would pass `Available` or `Unavailable` as applicable.

Reply:
Aha, now I understand what this function is for. I wasn't aware this was for the command line feature. Makes sense to configure it as an enum now.

Reply:
@varungandhi-apple I updated the branch and fixed the bug with the macro. With the enum, I was also able to clean up some of the logic as well 👍 

I just don't understand what you meant with `printAvailabilitySpecSource` and `dumpAvailabilitySpecSource` though, I saw some types with similar methods but I don't enough about the compiler to know what they are used for. Could you clarify that?

Reply:
I suggested adding them to help with debugging, since C++ doesn't have automatic printing facilities like Swift does.